File: src/ImageCompression/Utils/Objects/DctConvertor.java
package ImageCompression.Utils.Objects;


import ImageCompression.Constants.State;
import ImageCompression.Constants.TypeQuantization;
import ImageCompression.Containers.Flag;
import ImageCompression.Utils.Functions.DCTMultiThread;
import org.jetbrains.annotations.NotNull;

/**
 * Class for transformation between DCT and Origin
 * use min Size and max Time
 * Created by Димка on 08.08.2016.
 */
public class DctConvertor {
    public enum State{DCT,ORIGIN}
    private State state ;

    private short[][] dataOrigin;
    private short[][] dataProcessed;

    private int Width, Height;
    private int duWidth, duHeight;

    private TypeQuantization tq;
    private Flag flag;

//    private boolean isReady=false;

    public DctConvertor(short[][] dataOrigin, State state, TypeQuantization tq, Flag flag) {
        this.dataOrigin = dataOrigin;
        dataProcessed=dataOrigin;//= new short[dataOrigin.length][dataOrigin[0].length];// = dataOrigin
        Width = dataOrigin.length;
        Height = dataOrigin[0].length;
        this.tq = tq;
        this.flag=flag;

        this.state=state;
        sizeCalculate();
    }
    private void sizeCalculate() {
        duWidth = Width / DCTMultiThread.SIZEOFBLOCK;
        duHeight = Height / DCTMultiThread.SIZEOFBLOCK;
        if (Width % DCTMultiThread.SIZEOFBLOCK != 0)
            duWidth++;
        if (Height % DCTMultiThread.SIZEOFBLOCK != 0)
            duHeight++;
     //   createMatrix();
    }

    /**
     * subtract the [0][0] element from each [%8][%8]
     */
    private void preProsses() {
        for (int i = 0; i < duWidth; i++) {
            for (int j = 0; j < duHeight; j++) {

                int curX = i * DCTMultiThread.SIZEOFBLOCK;
                int curY = j * DCTMultiThread.SIZEOFBLOCK;
                if(i!=0&&j!=0)
                    dataOrigin[curX][curY]=(short)(dataOrigin[0][0]-dataOrigin[curX][curY]);
            }
        }
    }

    /**
     * copy 8x8 matrix from dataOrigin started at [i][j] into buffer
     * @param buffer - target matrix to copy
     * @return buffer
     */
    private short[][] fillBufferForDU(int i,int j,@NotNull short[][]buffer){
        for (int x = 0; x < DCTMultiThread.SIZEOFBLOCK; x++) {
            for (int y = 0; y < DCTMultiThread.SIZEOFBLOCK; y++) {
                short value = 0;
                int curX = i * DCTMultiThread.SIZEOFBLOCK + x;
                int curY = j * DCTMultiThread.SIZEOFBLOCK + y;
                if (curX< Width && curY < Height)
                    value = dataOrigin[curX][curY];
                buffer[x][y] = value;
                // DU[i][j].setValue(val,x,y);
            }
        }
        return buffer;
    }

    @FunctionalInterface
    interface FIConvertor{
        short[][] convert(short[][] buf);
    }
    private short[][] directDCT(short[][] buf){
        if(flag.isChecked(Flag.Parameter.Alignment))
            minus128(buf);

        buf = DCTMultiThread.directDCT(buf);

        if(flag.getQuantization()== Flag.QuantizationState.First)
            DCTMultiThread.directQuantization(tq,buf);
        return buf;
    }
    private short[][] reverceDCT(short[][] buf){
        if(flag.getQuantization()== Flag.QuantizationState.First)
            DCTMultiThread.reverseQuantization(tq,buf);

        buf = DCTMultiThread.reverseDCT(buf);

        if(flag.isChecked(Flag.Parameter.Alignment))
            plus128(buf);

        return buf;
    }

    /**
     * set 8x8 matrix from buffer into dataProcessed started at [i][j]
     * @param buffer - matrix with information
     */
    private void fillDateProcessed(int i,int j,@NotNull short[][]buffer){
        for (int x = 0; x < DCTMultiThread.SIZEOFBLOCK; x++) {
            for (int y = 0; y < DCTMultiThread.SIZEOFBLOCK; y++) {

                int curX = i * DCTMultiThread.SIZEOFBLOCK + x;
                int curY = j * DCTMultiThread.SIZEOFBLOCK + y;
                if (curX< Width && curY < Height)
                    dataProcessed[curX][curY] = buffer[x][y];
            }
        }
    }

    /**
     * do transmormation between DCT and Origin states
     */
    private void dataProcessing() {
        short[][] buf = new short[DCTMultiThread.SIZEOFBLOCK][DCTMultiThread.SIZEOFBLOCK];
        if(state==State.DCT)
            preProsses();

        FIConvertor convertor=(state==State.ORIGIN)?this::directDCT:this::reverceDCT;

        for (int i = 0; i < duWidth; i++) {
            for (int j = 0; j < duHeight; j++) {
                buf=fillBufferForDU(i,j,buf);
                buf=convertor.convert(buf);
                fillDateProcessed(i,j,buf);
            }
        }
        if(state==State.ORIGIN)
            preProsses();


        if(state==State.ORIGIN)
            state=State.DCT;
        else if(state==State.DCT)
            state=State.ORIGIN;
//        isReady=true;
    }

    private void minus128(short [][] arr){
        for(int i=0;i<arr.length;i++) {
            for(int j=0;j<arr[0].length;j++){
                arr[i][j]-=128;
            }
        }
    }
    private void plus128(short [][] arr){
        for(int i=0;i<arr.length;i++) {
            for(int j=0;j<arr[0].length;j++){
                arr[i][j]+=128;
            }
        }
    }

    public State getState() {
        return state;
    }

    /**
     * Do main calculation if need
     * @return matrix with original date
     */
    public short[][] getMatrixOrigin() {
//        if(!isReady&&state==State.DCT)
//            dataProcessing();

        if(state==State.DCT)
            dataProcessing();

        return dataProcessed;
    }

    /**
     * Do main calculation if need
     * @return matrix with DCT date
     */
    public short[][] getMatrixDct() {
//        if(!isReady&&state==State.ORIGIN)
//            dataProcessing();

        if(state==State.ORIGIN)
            dataProcessing();

        return dataProcessed;
    }
}

File: ImageCompressionLib/src/ImageCompressionLib/Utils/Objects/DctConvertor.kt
package ImageCompressionLib.Utils.Objects


import ImageCompressionLib.Constants.SIZEOFBLOCK
import ImageCompressionLib.Constants.TypeQuantization
import ImageCompressionLib.Containers.Matrix.Matrix
import ImageCompressionLib.Containers.Matrix.ShortMatrix
import ImageCompressionLib.Containers.Parameters
import ImageCompressionLib.Containers.Type.Flag
import ImageCompressionLib.Utils.Functions.Dct.DctAlgorithm8x8
import ImageCompressionLib.Utils.Functions.Dct.DctUniversalAlgorithm

/**
 * Class for transformation between DCT and Origin
 * use min Size and max Time
 * Created by Димка on 08.08.2016.
 */
class DctConvertor(private val dataOrigin: Matrix<Short>, state: State, private val tq: TypeQuantization, private val parameters:Parameters,dctUtil:DctUniversalAlgorithm) {
    //    private boolean isReady=false;

    var state: State? = null
        private set
    private val dataProcessed: Matrix<Short>

    private val Width: Int
    private val Height: Int
    private var duWidth: Int = 0
    private var duHeight: Int = 0

    /**
     * Do main calculation if need
     * @return matrix with original date
     */
    fun getMatrixOrigin(): Matrix<Short>{
            if (state == State.DCT)
                dataProcessing()
            return dataProcessed
        }

    /**
     * Do main calculation if need
     * @return matrix with DCT date
     */
    fun getMatrixDct(): Matrix<Short>{

            if (state == State.ORIGIN)
                dataProcessing()

            return dataProcessed
        }

    enum class State {
        DCT, ORIGIN
    }

    init {
        dataProcessed = dataOrigin//= new short[dataOrigin.length][dataOrigin[0].length];// = dataOrigin
        Width = dataOrigin.width
        Height = dataOrigin.height

        this.state = state
        sizeCalculate()
    }

    private fun sizeCalculate() {
        duWidth = Width / DctAlgorithm8x8.SIZEOFBLOCK
        duHeight = Height / DctAlgorithm8x8.SIZEOFBLOCK
        if (Width % DctAlgorithm8x8.SIZEOFBLOCK != 0)
            duWidth++
        if (Height % DctAlgorithm8x8.SIZEOFBLOCK != 0)
            duHeight++
        //   createMatrix();
    }

    /**
     * subtract the [0][0] element from each [%8][%8]
     */
    private fun preProsses() {
        for (i in 0 until duWidth) {
            for (j in 0 until duHeight) {

                val curX = i * DctAlgorithm8x8.SIZEOFBLOCK
                val curY = j * DctAlgorithm8x8.SIZEOFBLOCK
                if (i != 0 && j != 0)
                    dataOrigin[curX,curY] = (dataOrigin[0,0] - dataOrigin[curX,curY]).toShort()
            }
        }
    }

    /**
     * copy 8x8 matrix from dataOrigin started at [i][j] into buffer
     * @param buffer - target matrix to copy
     * @return buffer
     */
    private fun fillBufferForDU(i: Int, j: Int, buffer: Matrix<Short>): Matrix<Short> {
        for (x in 0 until DctAlgorithm8x8.SIZEOFBLOCK) {
            for (y in 0 until DctAlgorithm8x8.SIZEOFBLOCK) {
                var value: Short = 0
                val curX = i * DctAlgorithm8x8.SIZEOFBLOCK + x
                val curY = j * DctAlgorithm8x8.SIZEOFBLOCK + y
                if (curX < Width && curY < Height)
                    value = dataOrigin[curX,curY]
                buffer[x,y] = value
                // DU[i][j].setValue(val,x,y);
            }
        }
        return buffer
    }


    internal interface FIConvertor {
        fun convert(buf: Matrix<Short>): Matrix<Short>
    }

    private fun directDCT(buf: Matrix<Short>): Matrix<Short>{
        var buf = buf
        if (flag.isChecked(Flag.Parameter.Alignment))
            minus128(buf)

        buf = DctAlgorithm8x8.directDCT(buf)

        if (flag.quantization == Flag.QuantizationState.First)
            DctAlgorithm8x8.directQuantization(tq, buf)
        return buf
    }

    private fun reverceDCT(buf: Matrix<Short>): Matrix<Short>{
        var buf = buf
        if (flag.quantization == Flag.QuantizationState.First)
            DctAlgorithm8x8.reverseQuantization(tq, buf)

        buf = DctAlgorithm8x8.reverseDCT(buf)

        if (flag.isChecked(Flag.Parameter.Alignment))
            plus128(buf)

        return buf
    }

    /**
     * set 8x8 matrix from buffer into dataProcessed started at [i][j]
     * @param buffer - matrix with information
     */
    private fun fillDateProcessed(i: Int, j: Int, buffer: Matrix<Short>) {
        for (x in 0 until DctAlgorithm8x8.SIZEOFBLOCK) {
            for (y in 0 until DctAlgorithm8x8.SIZEOFBLOCK) {

                val curX = i * DctAlgorithm8x8.SIZEOFBLOCK + x
                val curY = j * DctAlgorithm8x8.SIZEOFBLOCK + y
                if (curX < Width && curY < Height)
                    dataProcessed[curX,curY] = buffer[x,y]
            }
        }
    }

    /**
     * do transmormation between DCT and Origin states
     */
    private fun dataProcessing() {
        var buf=ShortMatrix(SIZEOFBLOCK, SIZEOFBLOCK).toMatrix()
        if (state == State.DCT)
            preProsses()

        val convertor = if (state == State.ORIGIN) this::directDCT  else this::reverceDCT

        for (i in 0 until duWidth) {
            for (j in 0 until duHeight) {
                buf = fillBufferForDU(i, j, buf)
                buf = convertor.invoke(buf)
                fillDateProcessed(i, j, buf)
            }
        }
        if (state == State.ORIGIN)
            preProsses()


        if (state == State.ORIGIN)
            state = State.DCT
        else if (state == State.DCT)
            state = State.ORIGIN
        //        isReady=true;
    }

    private fun minus128(arr: Matrix<Short>) {
        for (i in 0 until arr.width) {
            for (j in 0 until arr.height){
                arr[i,j] =(arr[i,j]- 128).toShort()
            }
        }
    }

    private fun plus128(arr: Matrix<Short>) {
        for (i in 0 until arr.width) {
            for (j in 0 until arr.height) {
                arr[i,j] =(arr[i,j]+128).toShort()
            }
        }
    }
}
